fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/EquationRewriter.java

```diff
@@ -0,0 +1,214 @@
+
+                if (equationMap.containsKey(scalarOperator)) {
+                    Optional<ColumnRefOperator> mappedColumnRef = equationMap.get(scalarOperator).stream().findFirst();
+                    if (!mappedColumnRef.isPresent()) {
+                        return null;
+                    }
+                    if (columnMapping != null) {
+                        ColumnRefOperator replaced = columnMapping.get(mappedColumnRef.get());
+                        if (replaced == null) {
+                            return null;
+                        }
+                        return replaced.clone();
+                    } else {
+                        return mappedColumnRef.get().clone();
+                    }
+                } else if (extendedEquationMap.containsKey(scalarOperator)) {
+
+                    Optional<Pair<ColumnRefOperator, ScalarOperator>> mappedColumnAndExprRef =
+                            extendedEquationMap.get(scalarOperator).stream().findFirst();
+
+                    if (!mappedColumnAndExprRef.isPresent()) {
+                        return null;
+                    }
+
+                    if (columnMapping != null) {
+                        ColumnRefOperator basedColumn = mappedColumnAndExprRef.get().first;
+                        ColumnRefOperator replaced = columnMapping.get(basedColumn);
+                        if (replaced == null) {
+                            return null;
+                        }
+                        ScalarOperator newExpr = mappedColumnAndExprRef.get().second.clone();
+                        return replaceColInExpr(newExpr, basedColumn,
+                                replaced.clone()) ? newExpr : null;
+                    } else {
+                        return mappedColumnAndExprRef.get().second.clone();
+                    }
+                }
+
+                return null;
+            }
+
+            private boolean replaceColInExpr(ScalarOperator expr, ColumnRefOperator oldCol, ScalarOperator newCol) {
+                for (int i = 0; i < expr.getChildren().size(); i++) {
+                    if (oldCol.equals(expr.getChild(i))) {
+                        expr.setChild(i, newCol);
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+        };
+        return expr.accept(shuttle, null);
+    }
+
+    public boolean containsKey(ScalarOperator scalarOperator) {
+        return equationMap.containsKey(scalarOperator);
+    }
+
+    public Collection<ColumnRefOperator> get(ScalarOperator scalarOperator) {
+        return equationMap.get(scalarOperator);
+    }
+
+    public void addMapping(ColumnRefOperator col, ScalarOperator expr) {
+        equationMap.put(expr, col);
+        Pair<ScalarOperator, ScalarOperator> extendedEntry = new EquationTransformer(expr, col).getMapping();
+        extendedEquationMap.put(extendedEntry.first, Pair.create(col, extendedEntry.second));
```

Comment:
is that possible to put `expr ` in either `equationMap ` or `extendedEquationMap `? Does not need to put both maps?

Reply:
if `expr` or `col` is changed after EquationTransformer, put it in `extendedEquationMap `, otherwise into `equationMap `?

Reply:
fixed